src/Enzyme.jl

```diff
@@ -1380,17 +1281,24 @@ of shape `size(output)` of values of the input type.
     else
         Val(codegen_world_age(Core.Typeof(f), tt))
     end
-    primal, adjoint = Enzyme.Compiler.thunk(opt_mi, FA, BatchDuplicatedNoNeed{rt}, tt′, #=Split=# Val(API.DEM_ReverseModeGradient), #=width=#Val(chunk), ModifiedBetween, #=ReturnPrimal=#Val(false), #=ShadowInit=#Val(false), RABI, Val(ErrIfFuncWritten))
-    
+    primal, adjoint = Enzyme.Compiler.thunk(opt_mi, FA, BatchDuplicatedNoNeed{rt}, tt′,
+                                            #=Split=# Val(API.DEM_ReverseModeGradient),
+                                            #=width=#Val(chunk), ModifiedBetween,
+                                            #=ReturnPrimal=#Val(false), #=ShadowInit=#Val(false),
+                                            RABI, Val(ErrIfFuncWritten))
     if num * chunk == n_out_val
         last_size = chunk
         primal2, adjoint2 = primal, adjoint
     else
         last_size = n_out_val - (num-1)*chunk
         tt′ = Tuple{BatchDuplicated{Core.Typeof(x), last_size}}
-        primal2, adjoint2 = Enzyme.Compiler.thunk(opt_mi, FA, BatchDuplicatedNoNeed{rt}, tt′, #=Split=# Val(API.DEM_ReverseModeGradient), #=width=#Val(last_size), ModifiedBetween, #=ReturnPrimal=#Val(false), #=ShadowInit=#Val(false), RABI, Val(ErrIfFuncWritten))
+        primal2, adjoint2 = Enzyme.Compiler.thunk(opt_mi, FA, BatchDuplicatedNoNeed{rt}, tt′,
+                                                  #=Split=# Val(API.DEM_ReverseModeGradient),
+                                                  #=width=#Val(last_size), ModifiedBetween,
+                                                  #=ReturnPrimal=#Val(false), #=ShadowInit=#Val(false),
+                                                  RABI, Val(ErrIfFuncWritten))
     end
-
+    #TODO: this is broken for static arrays
```

Comment:
yeah this function presently assumes that the returned array is mutable, which perhaps should be added to the docstring (and later remedied).

Reply:
Yeah this was somewhat a note for myself as I definitely intend to make versions of this for static arrays at some point.  This is part of why I think it would be nice to have more autodiff methods, we could get it working there once and for all and not worry about separate `gradient` methods breaking it.

Reply:
I guess I'm not sure what change to autodiff would enable that? (i.e. you can use autodiff to implement this, the difference being for mutable (e.g. duplicated) vs immutable (e.g. active) data you need to pass it as an argument vs get it as a result). The reason is described a bit here: https://enzyme.mit.edu/julia/stable/faq/#Mixed-activity

But basically (and feel free to extend docs to explain this more if unclear), suppose you have a square function you want to differentiate, that takes a float

```julia
function grad_square(x, dx)
    dx += 2 * x
    return nothing
end
x = 3.14
dx = make_zero(x)
grad_square(x, dx)
@show dx # 0.0, not the intended 2 * 3.14
```

if it's a float dx is passed by value so we won't see it changed in the outer scope.

MixedDuplicated is a funky combo of the two that we've implemented to support code in either convention (including a struct of a float and an array), though at a slight perf loss.

Reply:
those docs should probably be updated now to say we have a MixedDuplicated type which means you don't need to change your original code to be ref-based, and you'd pass `MixedDuplicated(x, Ref(dx))` as an input.